### ISSUE TITLE
IndexedReferenceFinder: possibly too wide scope of searching

### ISSUE THREAD

**Issue by przepompownia** [phpactor/phpactor#2644]
For example, go to `lib/CodeTransform/Adapter/WorseReflection/Transformer/ImplementContracts.php` and find references to `$reflector` private property. 

Currently as a user I get notifications about scanning ~300 references
![image](https://github.com/phpactor/phpactor/assets/11404453/1b1d83b8-54c0-4b8e-9ecd-17b09e497f4a)

I started to debug what happens. 
1. `$reflector` is detected as a variable, not property. It probably needs some fixing.
2. even assuming that we search for references to a variable I need an explanation why searching is performed outside the class. Not sure if it's worth attention after fixing 1.

**Comment by mamazu** [comment]
I mean for properties it also would make sense to take the visibility into account. Like searching for a private variable outside of the class makes no sense and searching for protected variables outside of the inheritance hierarchy makes no sense as well.

**Comment by przepompownia** [comment]
For the order in `class X { function __construct(public $<>y) { print($y); } }` `$y` needs to be considered both property and variable.

**Comment by mamazu** [comment]
Yeah, but both should have a very limited scope. The variable only in the constructor and the member only in this class.

**Comment by dantleech** [comment]
Fwiw the legacy reference finder does take visibility into account and is obviously many times faster for private and protected properties, it' a missing feature in the "new" system

**Comment by przepompownia** [comment]
At the moment I give up. I found that we search for `property` type (as expected) in MemberQuery 

![image](https://github.com/phpactor/phpactor/assets/11404453/56c06d80-4564-4565-95e2-44dfcb40f381)

but have not idea when and how the list of references should be restricted.

**Comment by mamazu** [comment]
Okay, I think that might be an indexer problem, it's looking through all references of all files. Which for the `$reflector` is 218 file references.

**Comment by mamazu** [comment]
@dantleech I have a question because it sounds so obvious:
* When indexing a member we only index (for reference finding) the references with file and offset
* When finding references: we open the file reflect on it and check if the class is in the class hierarchy of our class.

Why don't we also index the context of the reference (for example the class name) into the record? Yes this increases the size of the index but it will drastically reduce the amount of work reference finding takes.

**Comment by dantleech** [comment]
> Yes this increases the size of the index but it will drastically reduce the amount of work reference finding takes

AFAIK we do index the class name (the `MemberRecord::$containerType`) but we cannot do this when it's not known at compile time. statically analysing the code when creating the index would make the indexing process 1000s of times slower (not to mention that the index can often be inaccurate for various reasons).

**Comment by dantleech** [comment]
> have not idea when and how the list of references should be restricted accordingly to the visibility.

basically you don't need the indexer here, you only need reflection. so refactoring is probably required.

**Comment by mamazu** [comment]
> AFAIK we do index the class name (the MemberRecord::$containerType) but we cannot do this when it's not known at compile time.

It looks like this `proprety` is null however. At least when I go to for References for `WorseReflectionNavigatorTest::reflector` then the memeber record it returns has no container type and non of the 100+ references have it. Maybe that's the problem.

**Comment by dantleech** [comment]
For private and protected the speed can be i increased by 10-1000+ times by not using the indexer at all... which is exactly what the non LSP version does

**Comment by przepompownia** [comment]
I checked (without digging into the code) if #2849 solves also this issue - rather still not:

![Image](https://github.com/user-attachments/assets/fd9a51d0-9ca6-4d43-82aa-d1874e441520)

**Comment by przepompownia** [comment]
...and tested #2856, which may be useful in some cases, but here local results are unavailable after first 10 seconds. I still have not checked what the order of searching depends on.
